### PR TITLE
Use dill to serialize logp functions in DensityDist

### LINKED ISSUE #3844: Parallel sampling hangs on macOS with Python 3.8 with pickling error
## Description of your problem

Starting with Python 3.8 the default start method for multiprocessing on macOS is `spawn` instead of `fork` - more info [here](https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods) and [here](https://bugs.python.org/issue33725). This means that some features that used to work will not continue to work on macOS. I specifically ran into this when using a `DensityDist` where the function was defined in place as follows because the function is not picklable:

```python
import pymc3 as pm

with pm.Model() as model:
    
    def val(x):
        return -0.5 * pm.math.sum(x ** 2)
    
    x = pm.Flat("x")
    pm.DensityDist("obs", val)
    
    pm.sample(chains=2, cores=2)
```

This fails with the following error that does not actually get caught. I was running in a Jupyter notebook and the notebook just hangs and this gets written to the command line:

```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/dforeman/research/projects/exoplanet-dev/case-studies/env/lib/python3.8/multiprocessing/spawn.py", line 116, in spawn_main
    exitcode = _main(fd, parent_sentinel)
  File "/Users/dforeman/research/projects/exoplanet-dev/case-studies/env/lib/python3.8/multiprocessing/spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
AttributeError: Can't get attribute 'val' on <module '__main__' (built-in)>
```

A temporary workaround that I found was to execute

```python
import multiprocessing as mp
mp.set_start_method("fork")
```

before importing PyMC3, but my understanding is that this won't be a long term solution. It doesn't seem like there's anything obvious to do about this, but I wanted to bring it up here in case folks had clever ideas!

## Versions and main components

* PyMC3 Version: 3.8
* Theano Version: 1.0.4
* Python Version: 3.8.1
* Operating system: macOS
* How did you install PyMC3: pip

### PR DESCRIPTION
**Pull request by aseyboldt**
Fix #3844 
When users pass functions to `pm.DensityDist`, then in most cases pickle will not be able to deal with those functions. The code will still work on Linux, because the default multiprocessing context is `fork`, and we don't even have to pickle the logp function. On Windows and Mac the default is `spawn` or `forkserver` however, and code using `pm.DensityDist` with a logp function that is not accessible in a module or protected by `if __name__ == '__main__` will fail. This includes all functions defined in a notebook as far as I understand.
We can work around this issue by manually serializing the logp function with dill.

_This PR introduces dill as a new required dependency to PyMC3, before it was optional._

Edit: I had some trouble with the cache of miniconda, so I also included a fix so that the cached environment is updated.

### CODE REVIEW

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ### Maintenance
 - Mentioned the way to do any random walk with `theano.tensor.cumsum()` in `GaussianRandomWalk` docstrings (see [#4048](https://github.com/pymc-devs/pymc3/pull/4048)).
 - Fixed numerical instability in ExGaussian's logp by preventing `logpow` from returning `-inf` (see [#4050](https://github.com/pymc-devs/pymc3/pull/4050)).
+- Use dill to serialize user defined logp functions in `DensityDist`. `dill` is now a required dependency. (see [#3844](https://github.com/pymc-devs/pymc3/issues/3844)).
```
Comment (twiecki) [inline RELEASE-NOTES.md]:
```suggestion
- Use dill to serialize user-defined logp functions in `DensityDist`. `dill` is now a required dependency. (see [#3844](https://github.com/pymc-devs/pymc3/issues/3844)).
```

File: scripts/create_testenv.sh [hunk 2]
```diff
@@ -24,13 +24,15 @@ ENVNAME="${ENVNAME:-testenv}"         # if no ENVNAME is specified, use testenv
 if [ -z ${GLOBAL} ]; then
   if conda env list | grep -q ${ENVNAME}; then
     echo "Environment ${ENVNAME} already exists, keeping up to date"
+    source activate ${ENVNAME}
```
Comment (twiecki) [inline scripts/create_testenv.sh]: I think new style is `conda activate` (or mamba).

File: RELEASE-NOTES.md (hunk 1 above)
Comment (twiecki) [reply RELEASE-NOTES.md]: I would add what this allows what wasn't possible before. did pickling just throw an error?

File: scripts/create_testenv.sh (hunk 2 above)
Comment (aseyboldt) [reply scripts/create_testenv.sh]: That doesn't work right now, but according to [this](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/use-conda-with-travis-ci.html#the-travis-yml-file) and [this](https://github.com/conda/conda/issues/7980) it should be fixed by sourcing the conda functions again:
```
source $CONDA_PREFIX/etc/profile.d/conda.sh
```
trying...

File: RELEASE-NOTES.md (hunk 1 above)
Comment (aseyboldt) [reply RELEASE-NOTES.md]: ok

File: scripts/create_testenv.sh (hunk 2 above)
Comment (twiecki) [reply scripts/create_testenv.sh]: If it's brittle probably best to keep it the way you had then.

Comment (aseyboldt) [reply scripts/create_testenv.sh]: Well, it is now the way the conda docs specify it...
But I can of course change it back.

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Thanks!

**Pull request merged**